components/fxa-client/README.md

```diff
@@ -0,0 +1,117 @@
+
+### Prerequisites
+
+To use this component, your application must be registered to [integrate with Firefox Accounts
+as an OAuth client](https://mozilla.github.io/ecosystem-platform/docs/process/integration-with-fxa)
+and have a unique OAuth `client_id`.
+
+You should also be familiar with how to integrate application-services components
+into an application on your target platform:
+* **Android**: integrate via the
+  [service-firefox-accounts](https://github.com/mozilla-mobile/android-components/tree/master/components/service/firefox-accounts/README.md)
+  component from android-components, which provides higher-level conveniences for state management, persistence,
+  and integration with other Android components.
+* **iOS**: start with the [guide to consuming rust components on
+  iOS](https://github.com/mozilla/application-services/blob/main/docs/howtos/consuming-rust-components-on-ios.md)
+  and take a look at the [higher-level Swift wrapper classes](./ios/FxAClient/).
+* **Other Platforms**: we don't know yet; please reach out on slack to discuss!
+
+### Core Concepts
+
+You should understand the core concepts of OAuth and the Firefox Accounts system
+before attempting to use this component. Please review the
+[Firefox Accounts Documentation](https://mozilla.github.io/ecosystem-platform/docs/features/firefox-accounts/fxa-overview)
+for more details.
+
+In particular, you should understand [the different types of auth token](
+https://github.com/mozilla/ecosystem-platform/pull/39)
+in the FxA ecosystem and how each is used, as well as how [OAuth scopes](
+https://github.com/mozilla/fxa/blob/main/packages/fxa-auth-server/docs/oauth/scopes.md)
+work for accessing related services.
+
+### Component API
+
+For details on how to use this component, consult the [public interface definition](./src/fxa_client.udl) or view the generated Rust API documentation by running:
+
+```
+cargo doc --no-deps --open
+```
+
+## Working on the Firefox Accounts client
+
+### Prerequisites
+
+To effectively work on the FxA Client component, you will need to be familiar with:
+
+* Our general [guidelines for contributors](../../docs/contributing.md).
+* The [core concepts](#core-concepts) for users of the component, outlined above.
+* The way we use [uniffi-rs](https://github.com/mozilla/uniffi-rs) to generate API wrappers for multiple languages, such as Kotlin and Swift.
+
+### Overview
+
+This component uses [uniffi-rs](https://github.com/mozilla/uniffi-rs) to create its
+public API surface in Rust, and then generate bindings for Kotlin and Swift. The
+code is organized as follows:
+
+* The public API surface is implemented in [`./src/lib.rs`](./src/lib/rs), with matching
+  declarations in [`./src/fxa_client.udl`](./src/fxa_client.udl) to define how it gets
+  exposed to other languages.
+    * All the implementation details are written in Rust and can be found under
+      [`./src/internal/`](./src/internal).
```

Comment:
I'm trying it out.

Broadly, I like having the contents of `src/lib.rs` mirror the public API exposed to foreign-language consumers, I think it makes it easier to form a consistent mental model of the component. Pushing all the non-public stuff into an `internal` crate makes that easy, but it might end up feeling a little forced.

For a crate like this, I like `internal` as a refactoring mechanism, basically push all the existing stuff in there and then we can incrementally clean it up. For crates built from the start to work with UniFFI, maybe it would be a bit pointless.

What do you think?

Reply:
> For a crate like this, I like `internal` as a refactoring mechanism, basically push all the existing stuff in there and then we can incrementally clean it up. For crates built from the start to work with UniFFI, maybe it would be a bit pointless.

It's a good idea in that context, but TBH I don't think it makes sense longer term (ie, it should be temporary) and that rust visibility is the right tool, particularly if fully leveraged (eg, `pub(super)` is probably underused by us).

In particular, I'm mildly concerned that pragmatic compromises might mean moving lots of code from `internal` just to expose it - or worse, marking code in `internal` as `pub`. For example, there are a number of things in places that's only `pub` because `places-utils.rs` wants to use it (and ditto autofill), which is unfortunate, but I tend to suck that up and just allow it to be marked `pub`)

That's weakly held though, particularly as it's not a "battle tested" opinion (although I'm trying to be better at it in autofill)

Reply:
Yep, I think I broadly agree with your point here - Rust has good mechanisms to manage what's public API and what's private, and we should lean on them rather than hack around them.

Apart from the contents of `internal/mod.rs`, most of what's in `internal` could quite happily live at the top level of the crate with limited visibility to prevent it from leaking out.

The goal is "don't have a bunch of `pub` stuff that's visible to Rust callers, but not available to foreign-language consumers" and I'm not too fussed about the implementation details. Which actually, that goal is pretty aligned with the recent email thread about Public API surface and mental models for our components...